File: frontend/src/services/notificationService.ts
import { apiFetch } from '@/lib/apiClient'

export interface NotificationResponse {
  id: number
  userId: number
  type: string
  title: string
  message: string
  link: string
  isRead: boolean
  createdAt: string
}

export interface UnreadCountResponse {
  count: number
}

export const notificationService = {
  getNotifications(userId: number) {
    return apiFetch<NotificationResponse[]>(`/api/notifications?userId=${userId}`)
  },
  getUnreadCount(userId: number) {
    return apiFetch<UnreadCountResponse>(`/api/notifications/unread-count?userId=${userId}`)
  },
  markAsRead(notificationId: number) {
    return apiFetch<void>(`/api/notifications/${notificationId}/read`, { method: 'PATCH' })
  },
  markAllAsRead(userId: number) {
    return apiFetch<void>(`/api/notifications/read-all?userId=${userId}`, { method: 'PATCH' })
  },
}

File: backend/src/main/java/com/eolmage/domain/notification/NotificationService.java
package com.eolmage.domain.notification;

import com.eolmage.domain.notification.dto.NotificationResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Transactional
    public void create(Long userId, String type, String title, String message, String link) {
        Notification notification = Notification.builder()
                .userId(userId)
                .type(type)
                .title(title)
                .message(message)
                .link(link)
                .build();
        notificationRepository.save(notification);
    }

    public List<NotificationResponse> getNotifications(Long userId) {
        return notificationRepository.findTop50ByUserIdOrderByCreatedAtDesc(userId)
                .stream()
                .map(NotificationResponse::from)
                .collect(Collectors.toList());
    }

    public long getUnreadCount(Long userId) {
        return notificationRepository.countByUserIdAndReadFalse(userId);
    }

    @Transactional
    public void markAsRead(Long notificationId) {
        notificationRepository.findById(notificationId)
                .ifPresent(Notification::markAsRead);
    }

    @Transactional
    public void markAllAsRead(Long userId) {
        notificationRepository.markAllAsReadByUserId(userId);
    }
}
